src/binaryReader.ts

```diff
@@ -0,0 +1,163 @@
+import { Readable } from 'stream';
+import { SmartBuffer, SmartBufferOptions } from 'smart-buffer';
+
+function createEnumChecker<T extends string, TEnumValue extends number>(
+  enumVariable: { [key in T]: TEnumValue }
+): (value: number) => value is TEnumValue {
+  // Create a set of valid enum number values.
+  const enumValues = Object.values<number>(enumVariable).filter(v => typeof v === 'number');
```

Comment:
`typeof NaN === 'number' // true` might this present a problem here?

Reply:
`NaN` is a valid typescript enum value. This works the same as `[1,NaN].includes(NaN) // true`

```ts
enum Color {
  Red = 1,
  Blue = NaN
}
const isColor = createEnumChecker(Color)(NaN); // true (is 'Color.Blue')
```